commons/src/main/java/org/eclipse/kapua/commons/service/internal/cache/KapuaCacheManager.java

```diff
@@ -97,39 +119,74 @@ private static URI getCacheConfig() {
+                    registeredCache.inc();
                     LOGGER.info("Created cache: {} - Expiry Policy: {} - TTL: {}", cacheName, EXPIRY_POLICY, TTL);
                 }
             }
         }
         return cache;
     }
 
+
+    private static void checkCacheManager() {
+        //called by synchronized section so no concurrency issues can arise
+        if (cacheManager == null) {
+            CachingProvider cachingProvider;
+            try {
+                if (!StringUtils.isEmpty(CACHING_PROVIDER_CLASS_NAME)) {
+                    cachingProvider = Caching.getCachingProvider(CACHING_PROVIDER_CLASS_NAME);
+                } else {
+                    cachingProvider = Caching.getCachingProvider();
+                }
+                //set the default cache flag
+                cacheStatus = 1;
+            } catch (CacheException e) {
+                //set the "default cache" flag (already done by initDefualtCacheProvider)
+                LOGGER.warn("Error while loading the CachingProvider... Loading the default one ({}).", DEFAULT_CACHING_PROVIDER_CLASS_NAME);
+                cachingProvider = initDefualtCacheProvider();
+            }
+            try {
+                cacheManager = cachingProvider.getCacheManager(CACHE_CONFIG_URI, null);
+            }
+            catch (Exception e) {
+                //anyway set the "default cache" flag (already done by initDefualtCacheProvider)
+                //second fallback
+                LOGGER.warn("Error while loading the CacheManager... Switching to CachingProvider default ({}). Error: {}", DEFAULT_CACHING_PROVIDER_CLASS_NAME, e.getMessage(), e);
+                cachingProvider = initDefualtCacheProvider();
+                cacheManager = cachingProvider.getCacheManager(CACHE_CONFIG_URI, null);
+            }
+        }
+    }
+
+    private static CachingProvider initDefualtCacheProvider() {
+        //set the default cache flag
+        cacheStatus = -1;
+        return Caching.getCachingProvider(DEFAULT_CACHING_PROVIDER_CLASS_NAME);
+    }
+
+    private static MutableConfiguration<Serializable, Serializable> initConfig() {
+        Factory expiryPolicyFactory;
+        if (ExpiryPolicy.TOUCHED.name().equals(EXPIRY_POLICY)) {
+            expiryPolicyFactory = TouchedExpiryPolicy.factoryOf(new Duration(TimeUnit.SECONDS, TTL));
+        } else {
+            expiryPolicyFactory = ModifiedExpiryPolicy.factoryOf(new Duration(TimeUnit.SECONDS, TTL));
+        }
+        MutableConfiguration<Serializable, Serializable> config = new MutableConfiguration<>();
+        config.setExpiryPolicyFactory(expiryPolicyFactory);
+        return config;
+    }
+
     /**
      * Utility method to cleanup the whole cache.
      */
     public static void invalidateAll() {
-        CACHE_MAP.forEach((cacheKey, cache) -> cache.clear());
+        CACHE_MAP.forEach((cacheKey, cache) -> {
+            cache.clear();
+            registeredCache.dec();
```

Comment:
I don't think registeredCache.dec() is enough, it should be set to zero, am I wrong ?

Reply:
My fault it's in the loop.